src/main/java/com/shipmentEvents/demo/deadlock5.java

```diff
@@ -0,0 +1,37 @@
+public class DeadlockExample {
+
+    private Object lock = new Object();
+    private boolean stopped = true;
+
+    public void run() {
+        synchronized (lock) {
+            isStopped();
+        }
+    }
+
+    private synchronized boolean isStopped() {
+        return stopped;
+    }
+
+    public synchronized void stopWorker() {
+        synchronized (lock) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 The code acquires two synchronization objects in **two different orders**, which can cause **circular wait** (deadlock).

 The **first order** is: first acquire `DeadlockExample.this` and then `lock`:

- enter `stopWorker()` at** line 16**

- **acquire synchronization** on `DeadlockExample.this` at entry to synchronized method `stopWorker()`, see** line 16**

- **acquire synchronization** on `lock` at** line 17**



 The **second order** is: first acquire `lock` and then `DeadlockExample.this`:

- enter `run()` at** line 6**

- **acquire synchronization** on `lock` at** line 7**

- call `isStopped()` at** line 8**

- **acquire synchronization** on `DeadlockExample.this` at entry to synchronized method `isStopped()`, see** line 12**



**Fix**  
 The fix is to acquire synchronization in only one order. Typically you can do this by adding extra synchronization to the synchronization order you want to prohibit.